### PR TITLE
Fix RVSDG->CFG translation

### PR DESCRIPTION
**Pull request by oflatt**
This PR completely rewrites RVSDG->CFG translation to a new, much nicer algorithm. The new algorithm returns a `TranslationResult`, representing a self-contained CFG with a start and end. The new algorithm is much easier to reason about.

Unfortunately, it adds a lot of unnecessary blocks because any node has to make new blocks. So we'll need an optimization pass in the future to fix those. Opening an issue

Also in the PR is a small fix to restructuring from @ezrosent, thank you!

### CODE REVIEW

**Comment by ezrosent** [comment]
High-level comment:

* This PR seems to "glue" two blocks together by doing a jump. That creates a lot of jumps even when there's no control flow in the original program.
* Is there a reason that we can't just concatenate the two basic blocks together when we merge them?

**Comment by oflatt** [comment]
Good points
I want to do both of those things as a separate pass after the fact. Yes, we could fix both those problems, but there would still be more cases of direct jumps to handle. See #66

File: src/rvsdg/to_cfg.rs [hunk 1]
```diff
@@ -27,54 +27,31 @@ use crate::{
...
 /// Represents the result of a RVSDG computation
 #[derive(Clone, Debug)]
 enum RvsdgValue {
     StateEdge,
     BrilValue(String, Type),
 }
 
+/// The return type of translating a part of an RVSDG
+/// Translating an RVSDG node always creates new blocks.
+/// These may be redundant or have a single input and output, but these
```
Comment (ezrosent) [inline src/rvsdg/to_cfg.rs:39]: Turn this into a TODO?

File: src/rvsdg/to_cfg.rs [hunk 2]
```diff
@@ -27,54 +27,31 @@ use crate::{
...
+/// are removed in an optimization pass.
+#[derive(Clone, Debug)]
+struct TranslationResult {
+    start: NodeIndex,
+    end: NodeIndex,
+    values: Vec<RvsdgValue>,
```
Comment (ezrosent) [inline src/rvsdg/to_cfg.rs:45]: I think it'd help to include an intuitive translation of what these fields mean, particularly `values`.

In particular, I think the RVSDGs that we currently generate have lots of variables plumbed through that are "redundant" for a given block; if we start hoisting those inputs out, is our use of `values` in this file still correct?

File: src/rvsdg/to_cfg.rs [hunk 3]
```diff
@@ -154,84 +131,157 @@ impl RvsdgFunction {
...
+                res2.start,
+                Branch {
+                    op: BranchOp::Jmp,
+                    pos: None,
+                },
+            );
+        }
+        TranslationResult {
+            start: results[0].start,
+            end: results[results.len() - 1].end,
```
Comment (ezrosent) [inline src/rvsdg/to_cfg.rs:201]: nit, optional: I think doing `results.last().unwrap()` is a bit easier to read than `results[results.len()-1]`.

File: src/rvsdg/to_cfg.rs [hunk 4]
```diff
@@ -154,84 +131,157 @@ impl RvsdgFunction {
...
+        }
+    }
+
+    /// Like [`SequenceResults`], but combines all of the results into
+    /// a multi-output [`TranslationResult`].
+    fn combine_results(&mut self, results: &[TranslationResult]) -> TranslationResult {
+        // first sequence results
+        let sequenced = self.sequence_results(results);
+        // now make a vec of all the values
+        let mut values = vec![];
```
Comment (ezrosent) [inline src/rvsdg/to_cfg.rs:212]: `Vec::with_capacity(results.len()) ?`

**Review by ezrosent — CHANGES_REQUESTED** [review]

File: src/rvsdg/to_cfg.rs (hunk 3 above)
Comment (oflatt) [reply src/rvsdg/to_cfg.rs:201]: done

File: src/rvsdg/to_cfg.rs (hunk 2 above)
Comment (oflatt) [reply src/rvsdg/to_cfg.rs:45]: Good idea
I added a comment. The idea is that `values` is a list of bindings for the outputs of the node. For single outputs, this is a vec of length one.
To answer your question, I don't see why that would be a problem. If an input is hoisted out, it's still a valid RVSDG and the node will have one less output.

File: src/rvsdg/to_cfg.rs (hunk 4 above)
Comment (oflatt) [reply src/rvsdg/to_cfg.rs]: changed to an iterator instead

**Review by ezrosent — APPROVED** [review]
sounds good! (though please fix the test failures)

**Comment by oflatt** [comment]
This is strange, I can't replicate the failure locally

**Comment by oflatt** [comment]
Nvm needed to merge main

**Comment by oflatt** [comment]
nevermind just needed to merge main

**Pull request merged**